rust/gui-client/src-tauri/src/client/gui/system_tray.rs

```diff
@@ -84,62 +77,107 @@ impl Tray {
             update_ready: state.release.is_some(),
         };
 
-        self.handle.set_tooltip(TOOLTIP)?;
-        self.handle.set_menu(build_app_state(state))?;
+        let menu = state.into_menu();
+        let menu_clone = menu.clone();
+        let app = self.app.clone();
+        let handle = self.handle.clone();
+
+        if Some(&menu) == self.last_menu_set.as_ref() {
+            tracing::debug!("Skipping redundant menu update");
+        } else {
+            self.app
+                .run_on_main_thread(move || {
+                    if let Err(error) = update(handle, &app, &menu) {
+                        tracing::error!(?error, "Error while updating tray icon");
+                    }
+                })
+                .unwrap();
+        }
         self.set_icon(new_icon)?;
+        self.last_menu_set = Some(menu_clone);
 
         Ok(())
     }
 
     // Only needed for the stress test
     // Otherwise it would be inlined
+    #[allow(clippy::unnecessary_wraps)]
     pub(crate) fn set_icon(&mut self, icon: Icon) -> Result<()> {
         if icon != self.last_icon_set {
             // Don't call `set_icon` too often. On Linux it writes a PNG to `/run/user/$UID/tao/tray-icon-*.png` every single time.
             // <https://github.com/tauri-apps/tao/blob/tao-v0.16.7/src/platform_impl/linux/system_tray.rs#L119>
             // Yes, even if you use `Icon::File` and tell Tauri that the icon is already
             // on disk.
-            self.handle.set_icon(icon_to_tauri_icon(&icon))?;
-            self.last_icon_set = icon;
+            let handle = self.handle.clone();
+            self.last_icon_set = icon.clone();
+            self.app
+                .run_on_main_thread(move || {
+                    handle.set_icon(Some(icon_to_tauri_icon(&icon))).unwrap();
+                })
+                .unwrap();
```

Comment:
I am assuming we are still going to deal with these `unwrap`s properly?

Reply:
One has to be there, removing the other one